proposals/2023/03-15_init-step.md

```diff
@@ -0,0 +1,325 @@
+# Add `InitStep` resource
+
+<!--
+The name of this markdown file should:
+
+1. Short and contain no more then 30 characters
+
+2. Contain the date of submission in MM-DD format
+
+3. Clearly state what the proposal is being submitted for
+-->
+
+| Key           | Value              |
+| :-----------: | :----------------: |
+| **Author(s)** | Jacob Floyd        |
+| **Reviewers** |                    |
+| **Date**      | March 15th, 2023   |
+| **Status**    | Waiting for Review |
+
+<!--
+If you're already working with someone, please add them to the proper author/reviewer category.
+
+If not, please leave the reviewer category empty and someone from the Vela team will assign it to themself.
+
+Here is a brief explanation of the different proposal statuses:
+
+1. Reviewed: The proposal is currently under review or has been reviewed.
+
+2. Accepted: The proposal has been accepted and is ready for implementation.
+
+3. In Progress: An accepted proposal is being implemented by actual work.
+
+NOTE: The design is subject to change during this phase.
+
+4. Cancelled: While or before implementation the proposal was cancelled.
+
+NOTE: This can happen for a multitude of reasons.
+
+5. Complete: This feature/change is implemented.
+-->
+
+## Background
+
+<!--
+This section is intended to describe the new feature, redesign or refactor.
+-->
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+<!--
+Provide your description here.
+-->
+
+Replace the injected "init" step (a pseudo-container) with `InitStep`, a new resource.
+
+Each `Build` will have one or more `InitSteps` that logically group related parts of the log data.
+
+An `InitStep` is a report by a "reporter" about a discrete part of the build setup. The "reporter" would be a logical part of the stack, like the "Pipeline Compiler" or the Runtime. For example, we could have these reporters be discrete initsteps (Reporter: name of the `InitStep`):
+- Pipeline Compiler: report info/debug logs
+- Docker Runtime: report network setup
+- Docker Runtime: report volume setup
+- Kubernetes Runtime: Pod YAML
+
+Similar to a `Step`, where the log data is stored in a separate `Log` struct/table, the `InitStep` is also associated with a `Log` that stores the actual log/report.
```

Comment:
Another enhancement for this proposal might be to:
Optionally associate an `InitStep` with a `Step` or `Service`.

Doing this would take the place of prepending the `$ ...` step command that is running. That command, and any other info like env vars, would go in the InitStep's Log allowing the Step's log to be saved without modification (except of course for any secret masking).

Reply:
Proposal updated with this idea.